### PR TITLE
Automatically resolve `to_shopify` mapping chain from oldest to current version

### PR DESCRIPTION
**Pull request by danielpgross**
### TL;DR

Changes how integration mappings are defined from older Shopify taxonomy versions.

Before:
* **In `data`**, old Shopify integration versions were defined relative to the _current/latest_ version (ex. 2022-01 is defined relative to 2025-01-unstable)
  * Pain point: every time we added a new `to_shopify` mapping, we needed to also add it individually _to each older version_.
* **In `dist`**, old Shopify integration versions were generated relative to the current version

After:
* **In `data`**, old Shopify integration versions are defined relative to the _immediately next_ version (ex. 2022-01 is defined relative to 2024-07)
* **In `dist`**, old Shopify integration versions are generated relative to the current version

### How it works

* When initially loading MappingRules, category IDs are no longer resolved to `Category` objects using the current taxonomy. Instead, the raw category ID is stored.
* After all mappings to Shopify have been loaded, the category IDs are resolved:
  * We start from the newest integration version and resolve its category IDs against the current taxonomy.
  * We then move backwards one version at a time, resolving each version against the immediate next one.
* Once this process is done, each integration version has had its category IDs resolved to `Category` objects, _taking into account any mappings that appear in newer versions._

### Breakdown of changes

The PR is big, but most of the changes are test fixtures. The actual implementation changes are mostly in:
* `IntegrationVersion#load_all_from_source`
* `IntegrationVersion#resolve_to_shopify_mappings`

The other notable change is extracting out one of the tests from `IntegrationVersionTest` into a dedicated integration test (`GenerateMappingsDistTest`). Instead of heavily relying on mocking like the old test, the new dedicated test compares generated file contents against fixtures. This approach is more straightforward and easier to maintain.

### CODE REVIEW

**Comment by danielpgross** [comment]
* **#517** <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/517?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/517?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by elsom25 — APPROVED** [review]
Such an important change. Thank you

File: dev/test/integration/generate_mappings_dist_test.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      Category.add(aa_1)
+      Category.add(aa_2)
+      Category.add(aa_3)
+    end
+
+    teardown do
+      FileUtils.remove_entry(@tmp_path)
+    end
+
+    test "generate_all_distributions generates all_mappings.json and distribution files for all integration versions" do
```
[the remark this answers is not in the document]
Comment (danielpgross) [reply dev/test/integration/generate_mappings_dist_test.rb:32]: Updated 👍

File: dev/test/integration/generate_mappings_dist_test.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      assert_file_contents_match "shopify/shopify_2020-01_to_shopify_2025-01.txt"
+      assert_file_contents_match "shopify/shopify_2021-01_to_shopify_2025-01.json"
+      assert_file_contents_match "shopify/shopify_2021-01_to_shopify_2025-01.txt"
+      assert_file_contents_match "shopify/shopify_2022-01_to_shopify_2025-01.json"
+      assert_file_contents_match "shopify/shopify_2022-01_to_shopify_2025-01.txt"
+      assert_file_contents_match "foocommerce/shopify_2025-01_to_foocommerce_1.0.0.json"
+      assert_file_contents_match "foocommerce/shopify_2025-01_to_foocommerce_1.0.0.txt"
+    end
+
+    def assert_file_contents_match(file_path)
```
[the remark this answers is not in the document]
Comment (danielpgross) [reply dev/test/integration/generate_mappings_dist_test.rb:51]: Yup makes sense

File: dev/test/models/integration_version_test.rb [hunk 3]
```diff
@@ -35,11 +42,33 @@ class IntegrationVersionTest < TestCase
...
+      output_mappings = integration_versions
+        .index_by { "#{_1.name}/#{_1.version}" }
+        .transform_values(&:to_shopify_mappings)
+
+      assert_equal "gid://shopify/TaxonomyCategory/aa", output_mappings["shopify/2020-01"].first.output_category.gid
+      # Defined as mapping to aa-1 in the 2020-01 version, resolved through to aa-3
+      assert_equal "gid://shopify/TaxonomyCategory/aa-3", output_mappings["shopify/2020-01"].second.output_category.gid
+      # Defined as mapping to aa-2 in the 2021-01 version, resolved through to aa-3
+      assert_equal "gid://shopify/TaxonomyCategory/aa-3", output_mappings["shopify/2021-01"].first.output_category.gid
+      assert_equal "gid://shopify/TaxonomyCategory/aa-3", output_mappings["shopify/2022-01"].first.output_category.gid
```
[the remark this answers is not in the document]
Comment (danielpgross) [reply dev/test/models/integration_version_test.rb:71]: Done 👍

**Comment by danielpgross** [comment]
### Merge activity

* **Jan 8, 4:41 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Shopify/product-taxonomy/517).
